src/blocknative.rs

```diff
@@ -0,0 +1,297 @@
+    max_fee_per_gas: f64,
+}
+
+#[derive(Debug, Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+struct BlockPrice {
+    estimated_prices: Vec<EstimatedPrice>,
+}
+
+#[derive(Debug, serde::Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+struct Response {
+    block_prices: Vec<BlockPrice>,
+}
+
+struct Request<T> {
+    transport: T,
+    header: http::header::HeaderMap,
+}
+
+impl<T: Transport> Request<T> {
+    async fn gas_price(&self) -> Result<Response> {
+        self.transport
+            .get_json(API_URI, self.header.clone())
+            .await
+            .context("failed to get blocknative gas price")
+    }
+}
+
+pub struct BlockNative {
+    cached_response: Arc<Mutex<CachedResponse<Response>>>,
+    handle: JoinHandle<()>,
+}
+
+impl Drop for BlockNative {
+    fn drop(&mut self) {
+        self.handle.abort();
+        self.cached_response = Default::default();
+    }
+}
+
+impl BlockNative {
+    pub async fn new<T: Transport + 'static>(
+        transport: T,
+        header: http::header::HeaderMap,
+    ) -> Result<Self> {
+        let cached_response: Arc<Mutex<CachedResponse<Response>>> = Default::default();
+        let cached_response_clone = cached_response.clone();
+
+        //send one request to initially populate the cached response
+        let request = Request { transport, header };
+        match request.gas_price().await {
+            Ok(response) => match cached_response_clone.lock() {
+                Ok(mut data) => {
+                    *data = CachedResponse {
+                        time: Instant::now(),
+                        data: Some(response),
+                    };
+                }
+                Err(e) => {
+                    tracing::warn!(?e, "failed to obtain lock on cached response");
+                    return Err(anyhow!("failed to obtain lock on cached response"));
+                }
+            },
+            Err(e) => {
+                tracing::warn!(?e, "failed to get initial response from blocknative");
+                return Err(anyhow!("failed to get initial response from blocknative"));
+            }
+        }
+
+        //spawn task for updating the cached response every RATE_LIMIT seconds
+        let handle = task::spawn(async move {
+            loop {
+                match request.gas_price().await {
+                    Ok(response) => match cached_response_clone.lock() {
+                        Ok(mut data) => {
+                            *data = CachedResponse {
+                                time: Instant::now(),
+                                data: Some(response),
+                            };
+                        }
+                        Err(e) => tracing::warn!(?e, "failed to obtain lock on cached response"),
+                    },
+                    Err(e) => tracing::warn!(?e, "failed to get response from blocknative"),
+                }
+                tokio::time::sleep(RATE_LIMIT).await;
```

Comment:
Nit: We can put this at the start of the loop because just before starting the task we have already fetched once.

Reply:
Makes sense. Fixed.